File: Compiler Design/lexical.cpp
#include <iostream>
#include <string>
#include <vector>
#include <regex>
#include <unordered_map>

struct Token {
    std::string type;
    std::string value;
    int line;
};

std::unordered_map<std::string, std::string> keywords = {
    {"if", "IF"},
    {"else", "ELSE"},
    {"while", "WHILE"},
    {"return", "RETURN"},
    {"def", "DEF"}
};

struct Rule {
    std::string type;
    std::regex pattern;
};

std::vector<Rule> rules = {
    {"FLOAT",  std::regex(R"(\d+\.\d+)")},
    {"INT",    std::regex(R"(\d+)")},
    {"STRING", std::regex(R"("(\\.|[^"\\])*"|'(\\.|[^'\\])*')")},
    {"ID",     std::regex(R"([A-Za-z_][A-Za-z0-9_]*)")},
    {"EQ",     std::regex(R"(==)")},
    {"NEQ",    std::regex(R"(!=)")},
    {"LE",     std::regex(R"(<=)")},
    {"GE",     std::regex(R"(>=)")},
    {"ASSIGN", std::regex(R"(=)")},
    {"PLUS",   std::regex(R"(\+)")},
    {"MINUS",  std::regex(R"(\-)")},
    {"TIMES",  std::regex(R"(\*)")},
    {"DIVIDE", std::regex(R"(/)")},
    {"LT",     std::regex(R"(<)")},
    {"GT",     std::regex(R"(>)")},
    {"LPAREN", std::regex(R"(\()")},
    {"RPAREN", std::regex(R"(\))")},
    {"LBRACE", std::regex(R"(\{)")},
    {"RBRACE", std::regex(R"(\})")},
    {"COMMA",  std::regex(R"(,)")},
    {"SEMI",   std::regex(R"(;)")},
    {"COLON",  std::regex(R"(:)")}
};

std::vector<Token> tokenize(const std::string &input) {
    std::vector<Token> tokens;
    size_t pos = 0;
    int line = 1;

    while (pos < input.size()) {
        // Skip whitespace
        if (isspace(input[pos])) {
            if (input[pos] == '\n') line++;
            pos++;
            continue;
        }

        if (input.compare(pos, 2, "//") == 0) {
            while (pos < input.size() && input[pos] != '\n') pos++;
            continue;
        }
        if (input.compare(pos, 2, "/*") == 0) {
            pos += 2;
            while (pos < input.size() && !(input[pos-1] == '*' && input[pos] == '/')) {
                if (input[pos] == '\n') line++;
                pos++;
            }
            pos++;
            continue;
        }

        bool matched = false;
        for (auto &rule : rules) {
            std::smatch match;
            std::string remaining = input.substr(pos);
            if (std::regex_search(remaining, match, rule.pattern, std::regex_constants::match_continuous)) {
                std::string val = match.str();
                std::string type = rule.type;

                // check if ID is keyword
                if (type == "ID" && keywords.count(val)) {
                    type = keywords[val];
                }

                tokens.push_back({type, val, line});
                pos += val.size();
                matched = true;
                break;
            }
        }

        if (!matched) {
            std::cerr << "Illegal character '" << input[pos] 
                      << "' at line " << line << "\n";
            pos++;
        }
    }

    return tokens;
}

int main() {
    std::string code = R"(
        // demo program
        def add(a, b) {
            return a + b;
        }
        x = 123;
        y = 3.14;
        s = "hello\nworld";
        if x >= 10 {
            y = y * 2;
        }
    )";

    auto tokens = tokenize(code);

    for (auto &tok : tokens) {
        std::cout << tok.type << "   " 
                  << tok.value << "   " 
                  << tok.line << "\n";
    }

    return 0;
}

File: Compiler Design/lexical.py
import re
import ply.lex as lex

tokens = (
    'ID',
    'INT',
    'FLOAT',
    'STRING',
    'PLUS', 'MINUS', 'TIMES', 'DIVIDE',
    'ASSIGN', 'EQ', 'NEQ', 'LT', 'GT', 'LE', 'GE',
    'LPAREN', 'RPAREN', 'LBRACE', 'RBRACE',
    'COMMA', 'SEMI', 'COLON'
)

reserved = {
    'if': 'IF',
    'else': 'ELSE',
    'while': 'WHILE',
    'return': 'RETURN',
    'def': 'DEF'
}

tokens = tokens + tuple(reserved.values())

t_PLUS    = r'\+'
t_MINUS   = r'-'
t_TIMES   = r'\*'
t_DIVIDE  = r'/'
t_EQ      = r'=='
t_NEQ     = r'!='
t_LE      = r'<='
t_GE      = r'>='
t_LT      = r'<'
t_GT      = r'>'
t_ASSIGN  = r'='
t_LPAREN  = r'\('
t_RPAREN  = r'\)'
t_LBRACE  = r'\{'
t_RBRACE  = r'\}'
t_COMMA   = r','
t_SEMI    = r';'
t_COLON   = r':'

def t_FLOAT(t):
    r'\d+\.\d+'
    t.value = float(t.value)
    return t

def t_INT(t):
    r'\d+'
    t.value = int(t.value)
    return t

def t_STRING(t):
    r'(\"([^\\\n]|(\\.))*?\")|(\'([^\\\n]|(\\.))*?\')'
    
    s = t.value[1:-1]
    t.value = bytes(s, "utf-8").decode("unicode_escape")
    return t

def t_ID(t):
    r'[A-Za-z_][A-Za-z0-9_]*'
    
    t.type = reserved.get(t.value, 'ID')
    return t


def t_COMMENT_singleline(t):
    r'//.*'
    pass  

def t_COMMENT_multiline(t):
    r'/\*([^*]|\*(?!/))*\*/'
    
    t.lexer.lineno += t.value.count('\n')
    pass


def t_newline(t):
    r'\n+'
    t.lexer.lineno += len(t.value)


t_ignore  = ' \t\r'


def t_error(t):
    print(f"Illegal character {t.value[0]!r} at line {t.lexer.lineno}")
    t.lexer.skip(1)

def build_lexer(**kwargs):
    return lex.lex(**kwargs)

if __name__ == '__main__':
    test_input = r'''
    // This is a single line comment
    def add(a, b) {
        return a + b;
    }

    /* multi-line
       comment */
    x = 123;
    y = 3.14;
    s = "hello\nworld";
    if x >= 10 {
        y = y * 2;
    }
    '''

    lexer = build_lexer()
    lexer.input(test_input)

    while True:
        tok = lexer.token()
        if not tok:
            break
        print(f"{tok.type:<12} {repr(tok.value):<20} {tok.lineno}")

